## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when R is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [60]:
type(my_thing)
len(my_thing)

NameError: name 'my_thing' is not defined

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [116]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H" "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"]

2. Make one long list containing all the symbols.  Find the length of this list.


In [117]:
all_symbols = lower_case + upper_case + punctuation

len(all_symbols)

63

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [161]:
import numpy as np

lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate((lower_case, upper_case, punctuation))

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [162]:
import pandas as pd

symbols_df = pd.DataFrame(all_symbols)

symbols_df.rename(columns = {0:"Symbol"})

Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]

## Part Two: Decoding the secret message.


In [180]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret = top_secret[0].to_numpy()

In [181]:
top_secret = top_secret + 14
top_secret = top_secret*18 - 257
top_secret = np.exp(top_secret)
top_secret = top_secret**2

top_secret

array([5.900e+01, 5.000e+01, 2.900e+01, 9.000e+00, 2.900e+01, 7.300e+01,
       1.600e+01, 3.400e+01, 2.700e+01, 3.200e+01, 3.400e+01, 3.300e+01,
       3.300e+01, 8.200e+01, 9.100e+01, 3.300e+01, 4.800e+01, 4.000e+01,
       9.900e+01, 7.900e+01, 5.000e+01, 4.900e+01, 1.070e+02, 6.700e+01,
       6.200e+01, 6.100e+01, 9.300e+01, 6.400e+01, 8.300e+01, 1.210e+02,
       1.010e+02, 7.900e+01, 8.800e+01, 7.300e+01, 8.900e+01, 1.330e+02,
       1.350e+02, 3.797e+03, 9.780e+02, 1.610e+02, 9.800e+01, 3.805e+03,
       1.350e+02, 7.130e+02, 4.140e+02, 1.170e+02, 3.815e+03, 9.700e+01,
       2.940e+02, 1.160e+02, 3.823e+03, 1.530e+02, 1.870e+02, 2.770e+02,
       1.140e+02, 2.560e+02, 1.390e+02, 3.837e+03, 1.990e+02, 3.160e+02,
       3.843e+03, 1.645e+03, 1.900e+02, 1.530e+02, 3.851e+03, 6.610e+02,
       1.350e+02, 1.400e+02, 1.630e+02, 4.236e+03, 3.863e+03, 8.730e+02,
       2.900e+02, 2.920e+02, 3.871e+03, 3.210e+02, 2.350e+02, 3.250e+02,
       5.190e+02, 7.850e+02, 3.883e+03, 6.930e+02, 

In [183]:
count = 0
for i in top_secret:
    if i < 17:
        count += 1
print(count)

352


In [195]:
matrix = top_secret.reshape(-1,5, order = "F")

print(matrix)

[[5.900e+01 1.440e+02 3.610e+02 9.000e+00 2.050e+01]
 [5.000e+01 1.000e+00 1.521e+03 7.500e+00 3.050e+01]
 [2.900e+01 5.290e+02 2.250e+02 1.050e+01 3.000e+00]
 [9.000e+00 3.610e+02 2.250e+02 3.500e+00 9.000e+00]
 [2.900e+01 3.721e+03 1.600e+01 4.000e+00 5.000e-01]
 [7.300e+01 1.521e+03 3.721e+03 3.050e+01 1.000e+00]
 [1.600e+01 6.400e+01 1.000e+00 5.000e-01 5.000e+00]
 [3.400e+01 1.000e+00 5.290e+02 7.000e+00 7.500e+00]
 [2.700e+01 1.521e+03 6.400e+01 2.000e+00 1.050e+01]
 [3.200e+01 3.721e+03 8.100e+01 3.050e+01 9.500e+00]
 [3.400e+01 9.000e+00 1.440e+02 1.950e+01 3.050e+01]
 [3.300e+01 1.000e+00 2.500e+01 4.000e+00 2.000e+00]
 [3.300e+01 1.521e+03 3.721e+03 9.000e+00 5.000e-01]
 [8.200e+01 9.000e+00 8.100e+01 7.500e+00 1.250e+01]
 [9.100e+01 6.400e+01 1.960e+02 1.050e+01 2.750e+01]
 [3.300e+01 3.025e+03 3.721e+03 3.500e+00 3.050e+01]
 [4.800e+01 3.721e+03 1.521e+03 4.000e+00 1.450e+01]
 [4.000e+01 7.840e+02 1.640e+02 3.050e+01 5.000e-01]
 [9.900e+01 2.500e+01 3.250e+02 3.050e+01 6.00

In [194]:
evens = np.arange(2,383, 2)

print(evens)

[  2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34  36
  38  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70  72
  74  76  78  80  82  84  86  88  90  92  94  96  98 100 102 104 106 108
 110 112 114 116 118 120 122 124 126 128 130 132 134 136 138 140 142 144
 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176 178 180
 182 184 186 188 190 192 194 196 198 200 202 204 206 208 210 212 214 216
 218 220 222 224 226 228 230 232 234 236 238 240 242 244 246 248 250 252
 254 256 258 260 262 264 266 268 270 272 274 276 278 280 282 284 286 288
 290 292 294 296 298 300 302 304 306 308 310 312 314 316 318 320 322 324
 326 328 330 332 334 336 338 340 342 344 346 348 350 352 354 356 358 360
 362 364 366 368 370 372 374 376 378 380 382]


In [198]:
matrix[:,0] -= evens
matrix[18:24,2] -= 100
matrix[:,3:5] *= 2

array = matrix.flatten()

print(array)

[ 5.500e+01  1.440e+02  3.610e+02  1.800e+01  4.100e+01  4.200e+01
  1.000e+00  1.521e+03  1.500e+01  6.100e+01  1.700e+01  5.290e+02
  2.250e+02  2.100e+01  6.000e+00 -7.000e+00  3.610e+02  2.250e+02
  7.000e+00  1.800e+01  9.000e+00  3.721e+03  1.600e+01  8.000e+00
  1.000e+00  4.900e+01  1.521e+03  3.721e+03  6.100e+01  2.000e+00
 -1.200e+01  6.400e+01  1.000e+00  1.000e+00  1.000e+01  2.000e+00
  1.000e+00  5.290e+02  1.400e+01  1.500e+01 -9.000e+00  1.521e+03
  6.400e+01  4.000e+00  2.100e+01 -8.000e+00  3.721e+03  8.100e+01
  6.100e+01  1.900e+01 -1.000e+01  9.000e+00  1.440e+02  3.900e+01
  6.100e+01 -1.500e+01  1.000e+00  2.500e+01  8.000e+00  4.000e+00
 -1.900e+01  1.521e+03  3.721e+03  1.800e+01  1.000e+00  2.600e+01
  9.000e+00  8.100e+01  1.500e+01  2.500e+01  3.100e+01  6.400e+01
  1.960e+02  2.100e+01  5.500e+01 -3.100e+01  3.025e+03  3.721e+03
  7.000e+00  6.100e+01 -2.000e+01  3.721e+03  1.521e+03  8.000e+00
  2.900e+01 -3.200e+01  7.840e+02  1.640e+02  6.100e+01  1.000

### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [199]:
''.join(map(str, symbols_df.loc[top_secret]['Symbol']))

KeyError: '[55.00000000000038, 42.0000000000018, 16.999999999999606, -7.000000000000043, 8.999999999999606, 49.00000000000169, -11.99999999999988, 2.0000000000001563, -8.99999999999866, -8.000000000001524, -9.999999999999844, -14.999999999999304, -18.999999999999304, 26.000000000001975, 30.999999999994984, -30.999999999999304, -20.00000000000248, -32.000000000000234, 22.99999999999622, -0.9999999999990905, -33.9999999999982, -39.000000000001904, 15.000000000003467, -29.0000000000027, -37.999999999998835, -43.0000000000007, -14.999999999998636, -47.99999999999929, -33.00000000000013, 1.0000000000056275, -23.000000000004363, -48.99999999999909, -43.999999999997456, -62.99999999999831, -50.99999999999925, -11.000000000005201, -13.000000000002501, 3645.000000000161, 822.000000000004, 0.9999999999980105, -65.99999999999802, 3636.9999999999104, -37.0000000000025, 536.9999999999766, 233.9999999999783, -67.00000000000475, 3627.0000000000778, -94.99999999999932, 97.99999999998846, -84.00000000000117, 3618.999999999989, -54.99999999999159, -24.99999999999514, 60.99999999999733, -106.00000000000232, 32.000000000003865, -89.00000000000227, 3605.00000000004, -37.00000000000108, 76.00000000000477, 3598.999999999863, 1396.9999999999093, -62.00000000000554, -102.99999999999159, 3591.0000000001805, 397.0000000000132, -133.0000000000025, -131.99999999999582, -112.99999999999929, 3956.000000000199, 3579.000000000159, 585.0000000000022, -2.0000000000067075, -3.9999999999921556, 3570.9999999998945, 16.999999999991473, -73.00000000000836, 13.000000000017792, 203.00000000002524, 464.9999999999711, 3558.999999999827, 365.00000000003865, -141.0000000000098, 156.0000000000117, -145.0000000000097, 3549.0000000000014, 1347.0000000000314, -112.00000000000153, -153.0000000000067, 3541.000000000122, -178.00000000000753, 41.00000000001944, 137.99999999999528, 36.99999999999466, -141.00000000000543, 33.00000000001603, 290.0000000000216, -170.99999999999477, 162.99999999997567, 2715.999999999842, 3519.000000000087, 3517.0000000000346, 3514.999999999793, 521.0000000000297, -13.999999999989598, -195.99999999999125, 3506.999999999913, 1305.000000000038, -153.99999999999096, -194.99999999999835, 3499.000000000191, -55.00000000001296, -0.9999999999794227, -58.999999999994884, -204.99999999998832, 3488.9999999998663, 90.00000000000784, -235.00000000001128, 1282.9999999999764, -176.0000000000038, 118.99999999997294, 3476.9999999999923, -21.000000000013074, 193.00000000000136, 1270.9999999999188, -203.00000000000813, 69.99999999997146, -255.0000000000038, -254.00000000000756, -235.00000000000955, 2547.0000000000864, 3457.0000000000546, 3097.999999999919, 516.0000000000034, -244.99999999998465, 257.0000000000032, -273.0000000000155, 47.99999999998829, -252.9999999999966, 3441.0000000001505, 1238.9999999999766, -219.99999999998442, -260.99999999998926, 3432.9999999999395, 1006.0000000000182, -291.0000000000012, -289.9999999999944, -291.99999999998937, -272.9999999999871, 24.000000000016144, 227.00000000003718, -78.9999999999925, -296.9999999999933, -187.00000000000642, 2605.999999999975, 3409.000000000097, -144.9999999999797, 309.00000000001125, 3402.999999999926, 41.00000000001137, -96.9999999999734, -128.0000000000217, 2698.9999999999277, 3393.00000000017, 3391.00000000023, 3388.99999999979, 1781.9999999999932, -272.0000000000061, -313.0000000000046, 3380.9999999999445, -242.00000000001387, -343.0000000000023, 183.00000000000273, 13.000000000027171, 3370.999999999814, 1168.9999999999152, -289.9999999999977, -355.0000000000068, 1163.0000000000675, 3360.999999999938, -358.0000000000043, -282.99999999997596, 1155.0000000000568, -343.00000000001296, 2545.9999999998126, 3348.9999999999213, 1147.0000000001078, -312.0000000000173, -352.9999999999848, 3340.999999999879, -373.00000000001836, 144.0000000000004, 529.0000000000075, 360.9999999999857, 3720.999999999915, 1521.0000000000566, 64.00000000000071, 8.999999999999957, 3025.000000000071, 783.9999999999756, 24.99999999999943, 323.99999999999824, 1295.9999999999975, 440.99999999998073, 4.000000000000014, 99.99999999999808, 80.99999999999923, 16.00000000000012, 2915.99999999997, 195.99999999999315, 2116.0000000000377, 35.99999999999997, 169.00000000000708, 224.99999999999383, 3363.999999999921, 1156.0000000000107, 121.00000000000563, 484.0000000000242, 256.00000000000387, 3968.9999999998086, 1443.9999999999482, 48.999999999998096, 576.0000000000038, 3844.0000000001446, 2024.9999999999347, 624.9999999999714, 728.9999999999897, 163.99999999999494, 225.0000000000178, 441.0000000000098, 48.99999999999403, 63.99999999999494, 1520.9999999999295, 2808.9999999998763, 60.9999999999993, 5.0000000000002265, 18.999999999999623, 14.999999999999794, 5.9999999999999964, 11.999999999999334, 12.999999999999531, 53.999999999999716, 28.999999999999606, 23.000000000000167, 7.999999999999589, 14.00000000000055, 6.999999999999864, 39.000000000000725, 18.00000000000097, 21.00000000000074, 27.00000000000134, 1.9999999999998896, 3.0000000000001634, 55.00000000000065, 40.999999999998586, 17.999999999999915, 15.000000000000618, 20.999999999999506, 7.000000000000249, 8.000000000000028, 61.00000000000265, 1.000000000000055, 13.999999999999728, 3.9999999999997793, 38.99999999999844, 46.000000000000334, 4.999999999999933, 21.999999999999265, 15.999999999999178, 11.999999999999993, 22.999999999998813, 19.00000000000067, 9.000000000000455, 2.9999999999999876, 11.000000000000234, 61.99999999999752, 54.99999999999743, 33.99999999999815, 6.000000000000327, 54.00000000000268, 13.000000000000247, 52.99999999999889, 57.99999999999921, 26.999999999999762, 36.00000000000194, 55.99999999999912, 29.0000000000012, 25.000000000000806, 41.000000000000846, 10.000000000000453, 56.999999999998636, 29.999999999999588, 63.99999999999695] not in index'